### PR TITLE
interfaces/builtin: distribute code of touching allInterfaces

### PR DESCRIPTION
**Pull request by zyga**
This patch moves code that registers an interface in allInterfaces, as
well as the code that ensures a given interface is registered into the
modules related to the actual interface. This should lessen the
conflicts around adding a new interface and simplify the task
conceptually.

The only remaining centralized elements are base declaration entries
(per jdstrand's request) and the list of implicit interfaces (which
could be merged into interfaces easily later).

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: interfaces/builtin/all.go [hunk 1]
```diff
@@ -20,101 +20,35 @@
...
-}
+var (
+	allInterfaces []interfaces.Interface
+	sorted        bool
+)
 
 // Interfaces returns all of the built-in interfaces.
 func Interfaces() []interfaces.Interface {
+	if !sorted {
+		sort.Sort(byIfaceName(allInterfaces))
```
Comment (mvo5) [inline interfaces/builtin/all.go]: Curious, why do we care about sorting?

File: interfaces/builtin/all.go [hunk 2]
```diff
@@ -20,101 +20,35 @@
...
+		sorted = true
+	}
 	return allInterfaces
 }
+
+// registerIface appends the given interface into the list of all known interfaces.
+func registerIface(iface interfaces.Interface) {
+	allInterfaces = append(allInterfaces, iface)
```
Comment (mvo5) [inline interfaces/builtin/all.go]: Should we care about duplication here? I.e. do not allow to register an interface twice?

File: interfaces/builtin/all.go (hunk 1 above)
Comment (zyga) [inline interfaces/builtin/all.go]: We don't care about sorting at all but I wanted to ensure that it's not changing this property. The old list used to be sorted so this one can be as well.

File: interfaces/builtin/all.go (hunk 2 above)
Comment (zyga) [inline interfaces/builtin/all.go]: Yes, I was thinking about it as well. Doing this branch has uncovered a number of smaller things that ought to be improved and I'm already preparing another one (and I have ideas for two more).

**Pull request merged**